packages/pipeline-services/src/migration/migrateV4/index.ts

```diff
@@ -14,15 +14,26 @@
  * limitations under the License.
  */
 
-function migrate(pipeline: any) {
-  for (const node of pipeline.pipelines[0].nodes) {
-    node.app_data = {
-      label: node.app_data.ui_data?.label ?? "",
-      component_parameters: node.app_data,
-      ui_data: node.app_data.ui_data ?? {},
-    };
-    delete node.app_data.component_parameters.ui_data;
+const migrateNodes = (pipeline: any, index: number) => {
+  for (const node of pipeline.pipelines[index].nodes) {
+    if (node.type === "super_node") {
+      const index = pipeline.pipelines.findIndex(
+        (p: any) => p.id === node.subflow_ref.pipeline_id_ref
+      );
+      migrateNodes(pipeline, index);
+    } else {
+      node.app_data = {
+        label: node.app_data.ui_data?.label ?? "",
+        component_parameters: node.app_data,
+        ui_data: node.app_data.ui_data ?? {},
+      };
+      delete node.app_data.component_parameters.ui_data;
+    }
```

Comment:
we should just iterate over all nodes. The recursion here seems unnecessary

```ts
for (const pipeline of pipeline.pipelines) {
  for (const node of pipeline.nodes) {
    node.app_data = {
      label: node.app_data.ui_data?.label ?? "",
      component_parameters: node.app_data,
      ui_data: node.app_data.ui_data ?? {},
    };
    delete node.app_data.component_parameters.ui_data;
  }
}
```

Reply:
Thanks, I didn't think of that solution, probably because I was already writing it recursively while I was still figuring out the json format. I'll make this update

Reply:
pushed